### ISSUE TITLE
Extracting Similarity/Distance Matrix and Labels

### ISSUE THREAD

**Issue by gsalfourn** [MaartenGr/BERTopic#292]
@Maarten,

This is not an issue, but just a question about how to extract the similarity matrix and labels in BERTopic if one wishes to. The only workaround,  I could think of, given my limited experience with Python, is the following, but not sure if you have a much simpler approach. 

My approach involves adapting the `visualize_heatmap` function and returning the `distance_matrix` and `new_labels`. See full code as used below:

```
def topic_similarity(topic_model,
                      topics: List[int] = None,
                      top_n_topics: int = None,
                      n_clusters: int = None):
  
    # select topic embeddings
    if topic_model.topic_embeddings is not None:
        embeddings = np.array(topic_model.topic_embeddings)
    else:
        embeddings = topic_model.c_tf_idf

    # select topics based on top_n and topics args
    if topics is not None:
        topics = list(topics)
    elif top_n_topics is not None:
        topics = sorted(topic_model.get_topic_freq().Topic.to_list()[1:top_n_topics + 1])
    else:
        topics = sorted(list(topic_model.get_topics().keys()))

    # order heatmap by similar clusters of topics
    if n_clusters:
        if n_clusters >= len(set(topics)):
            raise ValueError("Make sure to set `n_clusters` lower than "
                             "the total number of unique topics.")

        embeddings = embeddings[[topic + 1 for topic in topics]]
        distance_matrix = cosine_similarity(embeddings)
        Z = linkage(distance_matrix, 'ward')
        clusters = fcluster(Z, t=n_clusters, criterion='maxclust')

        # Extract new order of topics
        mapping = {cluster: [] for cluster in clusters}
        for topic, cluster in zip(topics, clusters):
            mapping[cluster].append(topic)
        mapping = [cluster for cluster in mapping.values()]
        sorted_topics = [topic for cluster in mapping for topic in cluster]
    else:
        sorted_topics = topics

    # select embeddings
    indices = np.array([topics.index(topic) for topic in sorted_topics])
    embeddings = embeddings[indices]
    distance_matrix = cosine_similarity(embeddings)

    # create nicer labels
    new_labels = [[[str(topic), None]] + topic_model.get_topic(topic) for topic in sorted_topics]
    new_labels = ["_".join([label[0] for label in labels[:4]]) for labels in new_labels]
    new_labels = [label if len(label) < 30 else label[:27] + "..." for label in new_labels]
    
    return distance_matrix, new_labels
```

**Comment by MaartenGr** [comment]
That is also a fine way of approaching this! I figured you did not want to include the topic number in the label but that I should not make such assumptions 😅. Just one note that I realize now, perhaps it would be best to extract the labels as follows instead:

```python
labels = (topic_model.get_topic_info().sort_values("Topic", ascending=True).Name)[1:]
```

The reason for this is that we cannot assume the topics to be ordered. This makes sure that they are in the right order.

**Comment by gsalfourn** [comment]
thanks.

**Comment by ma-ji** [comment]
Marking this thread!

**Comment by dilaraadem** [comment]
Hi @MaartenGr , 
I have a question related to this issue. I want to apply Soft Cosine Measure instead of Cosine to compare two topic models. I was using `model.topic_embedding` for Cosine, however Soft Cosine is a bit trickier and doesn't have a simple one line code.
I tried adapting Gensim's implementation, but their way (see [here](https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/soft_cosine_tutorial.ipynb)) seemed like it creates extra work by using documents instead of embeddings, since they are comparing documents. 
Do you know if it is possible to apply Soft Cosine to `topic_embeddings `or `c_tf_idf `parameters (or any other that I am not aware of) of BERTopic? I am trying to compare list of topics, not documents. So I wouldn't prefer to go through the process of creating a dictionary in BoW format and then creating a tfidf, etc if there is already a parameter that comes with loading the model.
Thanks!

**Comment by MaartenGr** [comment]
@dilaraadem I haven't tried out soft cosine yet but it seems that it does require some tweaking to get it to work. Here, I believe you could replace documents with topics and apply the same structure. Do note that you can use `topic_model.vectorizer_model` to generate the bag-of-word representations and `topic_model.transformer` to generate the c-TF-IDF representations.

**Comment by rubypnchl** [comment]
> I think it would require less work if you approach it something like this:
> 
> ```python
> import numpy as np
> from sklearn.metrics.pairwise import cosine_similarity
> 
> distance_matrix = cosine_similarity(np.array(topic_model.topic_embeddings)[1:, :])
> labels = ["_".join(label.split("_")[1:]) for label in topic_model.get_topic_info().Name[1:]]
> ```
> 
> I purposefully ignore the -1 cluster as it is not an actual topic. Each index in the `distance_matrix` represents a label in `labels`.

Hi MaartenGr,

I am enjoying bertopic and doing different experiments with it.  Now, I want to find similarities between topics (For e.g., Similarity between T1,T2...Tn) generated in a dataset. I have tried the above code but I am getting errors, also I want to set a threshold for similar topics (any suggestions, I am assuming two topics if they have similarity greater than 70-80% are similar ).

import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd

embeddings = embeddings[[topic + topic_model._outliers for topic in topics]]
#print(embeddings)
distance_matrix = cosine_similarity(np.array(topic_model.topic_embeddings)[1:, :])
df = pd.DataFrame(distance_matrix)
labels = ["_".join(label.split("_")[1:]) for label in topic_model.get_topic_info().Name[1:]]
print(df)

but I am getting the following error:
IndexError                                Traceback (most recent call last)
Input In [223], in <cell line: 7>()
      5 embeddings = embeddings[[topic + topic_model._outliers for topic in topics]]
      6 #print(embeddings)
----> 7 distance_matrix = cosine_similarity(np.array(topic_model.topic_embeddings)[1:, :])
      8 df = pd.DataFrame(distance_matrix)
      9 labels = ["_".join(label.split("_")[1:]) for label in topic_model.get_topic_info().Name[1:]]

IndexError: too many indices for array: array is 0-dimensional, but 2 were indexed

I am too exhausted to find a good similarity calculation method. Kindly help me out!

Thank you,

**Comment by MaartenGr** [comment]
@rubypnchl In the most recent release (v0.12), the following is working for me:

```python
import pandas as pd

sim_matrix = cosine_similarity(topic_model.c_tf_idf_)
df = pd.DataFrame(sim_matrix, columns=topic_model.topic_labels_.values(), index=topic_model.topic_labels_.values())
```

That way, it generates a similarity matrix using the c-TF-IDF representations and puts it in a dataframe for easy access.

**Comment by chav-ngvyen** [comment]
Dear Maarten and everyone, thank you for starting this thread. Maarten, let's say that I have trained 2 BERTopic models and would like to find similar topics between them.  Conceptually, could you help me understand the difference between:

1) calculating the pairwise cosine similarity between each topic_model.topic_embeddings_ as you showed in [Finding similar topics between models](https://maartengr.github.io/BERTopic/getting_started/tips_and_tricks/tips_and_tricks.html#finding-similar-topics-between-models]
vs
2) calculating the pairwise cosine similarity between each topic_model.c_tf_idf as you showed in this example before? https://github.com/MaartenGr/BERTopic/issues/435

Thank you so much for creating an awesome package and for always being willing to help!

**Comment by MaartenGr** [comment]
They are actually quite similar to one another. The topic embeddings are the weighted average of the top n word embeddings in a topic by their c-TF-IDF score. In other words, we simply take the top n words from a c-TF-IDF matrix, extract word embeddings from them and then average them together weighted by their c-TF-IDF score. It tries to mimic the c-TF-IDF matrix as much as possible, which is likely to be changed in the upcoming release. 

In other words, both c-TF-IDF and topic embeddings are representations of topics, the former in a bag-of-words format and the latter an embedding-based representation.

**Comment by chav-ngvyen** [comment]
>  which is likely to be changed in the upcoming release.
Thank you so much for the explanation, Maarten. I will keep an eye out for the changes in the upcoming release as well! The package keeps getting better and better :)

**Comment by rubypnchl** [comment]
Dear Maarten,

I am trying to find similar topics between two topic models using the same embedding method, last time (September 2022) when I tested it worked fine for me but now, I am getting something irrelevant and I doubt the wrong cosine similarity.

Following is the code I am using to find similar topics with similarity more than 95%.

from sklearn.metrics.pairwise import cosine_similarity
sim_matrix = cosine_similarity(topic_model.topic_embeddings_, topic_model1.topic_embeddings_)
for i in range(0,len(sim_matrix)-1):
    for j in range(0,len(sim_matrix)-1):
        if sim_matrix[i][j]>0.90 and sim_matrix[i][j]<=0.999:
              #print topics with similarity greater than 0.90 threshold
            print(i, j, [i[0] for i in topic_model.get_topic(i)], [j[0] for j in topic_model1.get_topic(j)], sim_matrix[i][j])
 

Here is the sample of output I am getting: 

1 23 ['xml', 'tables', 'documents', 'table', 'taxonomy', 'document', 'notations', 'layout', 'syntax', 'layouts', 'documentation', 'scientific', 'schema', 'web', 'classify', 'sections', 'sharing', 'elements', 'hierarchical', 'universe'] ['housing', 'ogun', 'nigeria', 'government', 'council', 'rent', 'paying', 'budget', 'councils', 'democratic', 'tax', 'democratization', 'electoral', 'provision', 'nigerian', 'list', 'deficit', 'exercise', 'public', 'reform'] 0.9839393322283956

2 4 ['multiobjective', 'optimization', 'pareto', 'variables', 'optimisation', 'evolutionary', 'algorithm', 'optimal', 'genetic', 'optimizing', 'geometries', 'swarm', 'topology', 'virus', 'algorithms', 'hyperbolic', 'surrogate', 'solutions', 'dimensionality', 'compact'] ['table', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', ''] 0.9762782886241512


Here as per the output, the keywords representing topics are not showing any similarity but the similarity is more than 98% in the first case, while in the second case it is more than 97%.

I don't understand why is this happening. Second, we can see in ['table', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', ''], there are only empty strings than is it possible to have 97% cosine similarity. 
Third, how can I remove such kind of empty strings from topics? Here the input to BERTopic is the sentences, not paragraphs.

Kindly help me out! Many Thanks in advance!

**Comment by SaraAmd** [comment]
> ```python
> labels = (topic_model.get_topic_info().sort_values("Topic", ascending=True).Name)[1:]
> ```

I want to get the predicted topic for each data in my dataset after training the topic model. How can I do that?

**Comment by SaraAmd** [comment]
> > I think it would require less work if you approach it something like this:
> > ```python
> > import numpy as np
> > from sklearn.metrics.pairwise import cosine_similarity
> > 
> > distance_matrix = cosine_similarity(np.array(topic_model.topic_embeddings)[1:, :])
> > labels = ["_".join(label.split("_")[1:]) for label in topic_model.get_topic_info().Name[1:]]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I purposefully ignore the -1 cluster as it is not an actual topic. Each index in the `distance_matrix` represents a label in `labels`.
> 
> Hi MaartenGr,
> 
> I am enjoying bertopic and doing different experiments with it. Now, I want to find similarities between topics (For e.g., Similarity between T1,T2...Tn) generated in a dataset. I have tried the above code but I am getting errors, also I want to set a threshold for similar topics (any suggestions, I am assuming two topics if they have similarity greater than 70-80% are similar ).
> 
> import numpy as np from sklearn.metrics.pairwise import cosine_similarity import pandas as pd
> 
> embeddings = embeddings[[topic + topic_model._outliers for topic in topics]] #print(embeddings) distance_matrix = cosine_similarity(np.array(topic_model.topic_embeddings)[1:, :]) df = pd.DataFrame(distance_matrix) labels = ["_".join(label.split("_")[1:]) for label in topic_model.get_topic_info().Name[1:]] print(df)
> 
> but I am getting the following error: IndexError Traceback (most recent call last) Input In [223], in <cell line: 7>() 5 embeddings = embeddings[[topic + topic_model._outliers for topic in topics]] 6 #print(embeddings) ----> 7 distance_matrix = cosine_similarity(np.array(topic_model.topic_embeddings)[1:, :]) 8 df = pd.DataFrame(distance_matrix) 9 labels = ["_".join(label.split("_")[1:]) for label in topic_model.get_topic_info().Name[1:]]
> 
> IndexError: too many indices for array: array is 0-dimensional, but 2 were indexed
> 
> I am too exhausted to find a good similarity calculation method. Kindly help me out!
> 
> Thank you,

I want to get the predicted topic for each data in my dataset after training the topic model. How can I do that?

**Comment by MaartenGr** [comment]
@SaraAmd 

> I want to get the predicted topic for each data in my dataset after training the topic model. How can I do that?

You can find them with `topic_model.topics_`.

> I want to get the predicted topic distribution for each data in my dataset after training the topic model. How can I do that?

If you have set `calculate_probabilities=True`, then you can find them with `topic_model.probabilities_`.

**Comment by SaraAmd** [comment]
Thank you so much for the response. It only returns the probabilities for a few worlds even if I don't set the top_n_words and nr_topics t any number. Is there a way to return all the probabilities for all the worlds in the vocabulary? I am asking because after doing topic modeling on my dataset I need to measure the KL divergence on topic distribution for each data in my dataset

**Comment by MaartenGr** [comment]
You can extract the topic-term matrices for each topic using `topic_model.c_tf_idf_`. The words are then extracted with `topic_model.vectorizer_model.get_feature_names()`. If you want to model the probabilities of words in a document to each topic, you can use [`.approximate_distribution`](https://maartengr.github.io/BERTopic/getting_started/distribution/distribution.html) for that.

**Comment by AskerNC** [comment]
> I think it would require less work if you approach it something like this:
> 
> import numpy as np
> from sklearn.metrics.pairwise import cosine_similarity
> 
> distance_matrix = cosine_similarity(np.array(topic_model.topic_embeddings)[1:, :])
> labels = ["_".join(label.split("_")[1:]) for label in topic_model.get_topic_info().Name[1:]]
> I purposefully ignore the -1 cluster as it is not an actual topic. Each index in the `distance_matrix` represents a label in `labels`.

Dear Maarten
First of all, thank you so much for this amazing package!

I have a question regarding your comment, which I think relates to me not really understanding what the embedding of the outlier topic represents.
If you use the .transform() method with hdbscan_model=Basecluster and calculate_probabilities=False, Bertopic will assign some documents to the outlier -1 topic. This seems a bit weird to me. Is the argument that the embedding of the outlier is a bit random, so if the document has the highest similarity with that, it is likely an outlier?
As I understand it, the topic embedding of the outlier is the centroid of all documents that were assigned as outliers during fitting, which presumably doesn't have any coherence or meaningful centroid. 

I guess what I'm asking is, is there any meaningful interpretation of the topic embedding of the outlier topic, and does it make sense to assign documents as outliers using cosine similarity with it?

**Comment by MaartenGr** [comment]
@AskerNC If there is a larger similarity to the outlier topic than any other, then I would indeed assume the document to be an outlier. This is especially true considering the centroid is not at a meaningful place and should therefore not easily be connected to a document. So although the topic embedding of the outlier does not necessarily have much of a meaning (other than "everything else" or "miscellaneous"), it would make sense (at least to me) to assign outliers to it.

**Comment by AskerNC** [comment]
@MaartenGr Thank you for your answer. My worry would be that the somewhat random outlier centroid is at an arbitrary point, so whether a document is determined to be an outlier is not determined by how similar it is to its most similar topic, but rather how similar it is, relative to some arbitrary point.
A different way of determining outliers I've considered would be to say: if the document's maximum similarity to its most similar topic is below some cutoff (e.g. 0.5), it is determined to be an outlier. 
But I take your point that, given some document is closer to the outlier topic than to an actual topic, it is likely an outlier
